src/cascadia/TerminalSettingsModel/Command.cpp

```diff
@@ -65,10 +68,158 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         return _subcommands ? _subcommands.GetView() : nullptr;
     }
 
+    // Function Description:
+    // - reports if the current command has nested commands
+    // - This CANNOT detect { "name": "foo", "commands": null }
     bool Command::HasNestedCommands() const
     {
         return _subcommands ? _subcommands.Size() > 0 : false;
     }
+
+    // Function Description:
+    // - reports if the current command IS a nested command
+    // - This CAN be used to detect cases like { "name": "foo", "commands": null }
+    bool Command::IsNestedCommand() const noexcept
+    {
+        return _nestedCommand;
+    }
+
+    bool Command::HasName() const noexcept
+    {
+        return _name.has_value();
+    }
+
+    hstring Command::Name() const noexcept
+    {
+        if (_name.has_value())
+        {
+            // name was explicitly set, return that value.
+            return hstring{ _name.value() };
+        }
+        else if (_ActionAndArgs)
+        {
+            // generate a name from our action
+            return get_self<implementation::ActionAndArgs>(_ActionAndArgs)->GenerateName();
+        }
+        else
+        {
+            // we have no name
+            return {};
+        }
+    }
+
+    void Command::Name(const hstring& value)
+    {
+        if (!_name.has_value() || _name.value() != value)
+        {
+            _name = value;
+        }
+    }
+
+    std::vector<Control::KeyChord> Command::KeyMappings() const noexcept
+    {
+        return _keyMappings;
+    }
+
+    // Function Description:
+    // - Add the key chord to the command's list of key mappings.
+    // - If the key chord was already registered, move it to the back
+    //   of the line, and dispatch a notification that Command::Keys changed.
+    // Arguments:
+    // - keys: the new key chord that we are registering this command to
+    // Return Value:
+    // - <none>
+    void Command::RegisterKey(const Control::KeyChord& keys)
+    {
+        if (!keys)
+        {
+            return;
+        }
+
+        // Check if we registered this key chord before
+        for (auto pos = _keyMappings.begin(); pos < _keyMappings.end(); ++pos)
+        {
+            if (keys.Modifiers() == pos->Modifiers() && keys.Vkey() == pos->Vkey())
+            {
+                // KeyChord was already registered.
+                if (*pos == _keyMappings.back())
+                {
+                    // It's already the latest key registered.
+                    return;
+                }
+                else
+                {
+                    // Move the new KeyChord to the back of the line.
+                    _keyMappings.erase(pos);
+                    _keyMappings.push_back(*pos);

```

Comment:
This code is incorrect and will crash, because `erase()` invalidates all iterators in the range `[pos, end())`.
--> You cannot copy `*pos`.

Fix:
```cpp
_keyMappings.erase(pos);
break;
```

Oh and: Since the prior if condition has a `return;` you can remove the `else` and unnest this by 1 (if you want to).

---

But the best improvement would be to just write it like that:
```cpp
void Command::RegisterKey(const Control::KeyChord& keys)
{
    if (!keys)
    {
        return;
    }

    EraseKey(keys);
	_keyMappings.emplace_back(keys);
}
```